### DISCUSSION TITLE
How to introduce BindingReducer() when not using var body: some ReducerProtocol<State, Action>

### DISCUSSION THREAD

**Discussion by arbyruns** [pointfreeco/swift-composable-architecture#1588] [Q&A]
We're currently behind in our structure in our project where we're setting our state and action like so

```
public enum AppState: Equatable...
public enum AppAction: Equatable...
public let appReducer = Reducer<AppState, AppAction, AppEnvironment>.combine(
...
    Reducer { state, action, env in 
    ...
    }
)
```

We're running into an issue where we're prompted with some issue and a solution of `To fix this, invoke "BindingReducer()" from your feature reducer's "body".`

I've tried adding `BindingReducer()` before the Reducer, but running into some errors and I'm not entirely sure this is the correct approach.

```
Cannot convert value of type 'BindingReducer<State, Action>' to expected argument type 'AnyReducer<AppState, AppAction, AppEnvironment>'
```

**Comment by mbrandonw** [comment]
Hi @arbyruns, the issue you are running into is that you are sending binding actions into the system but you haven't hooked up your reducer with the binding logic. So no changes are going to actually be made to your state.


Unfortunately that warning message describes the solution for the new `ReducerProtocol` style, but since you are using the old `Reducer` struct style the fix is a little different. We should probably update that message.

The fix is for you to use the `.binding()` method on the reducer that is using bindable state and actions:

```swift
Reducer { state, action, environment in 
}
.binding()
```

Does that fix the warning?

**Reply by arbyruns** [reply]
Thank you for the quick reply.  I think it get's me closer, but now I'm getting 

```
Referencing instance method 'binding()' on 'AnyReducer' requires that 'AppAction' conform to 'BindableAction'
```

**Reply by mbrandonw** [reply]
You should apply `.binding()` to the reducer that is using `BindableAction`. Which reducer is that?

**Reply by arbyruns** [reply]
Ah I see. Let me dig and see. Thank you!

**Reply by arbyruns** [reply]
So I think I have it in the right place, but still seeing the error. 

```
 A binding action sent from a view store at "Login/EmailRegistrationView.swift:93" was not handled. …

  Action:
    EmailRegistrationAction.binding(.set(_, SomeTextHere))

To fix this, invoke "BindingReducer()" from your feature reducer's "body".
```
```
public let emailRegistrationReducer = Reducer<
    EmailRegistrationState,
    EmailRegistrationAction,
    EmailRegistrationEnvironment
> { state, action, env in
    switch action {
                 .... 
                       }
   }
.binding()
```

**Reply by mbrandonw** [reply]
You may need to share more code to figure out the problem.

FWIW, this test passes with no warnings and demonstrates how to use bindable state:

https://github.com/pointfreeco/swift-composable-architecture/blob/5b78fbcb0583568392762b15a262b3106cfb5185/Tests/ComposableArchitectureTests/BindingTests.swift

Does this example differ from your code at all?

**Reply by arbyruns** [reply]
Thanks, let me circle back tomorrow since it's getting late where I'm at.

**Reply by arbyruns** [reply]
Sorry getting back to this now. So it does look similar. Here's some more that may help. Could one of our pullbacks be contributing to this issue?

```
public enum EmailRegistrationAction: Equatable, BindableAction {
    case binding(BindingAction<EmailRegistrationState>)
    case joinNowTapped
    case logInTapped
    case passwordHideToggleTapped
    case privacyPolicyTapped
    case receivedRegistrationResponse(Result<LoginRequest, ApiError>)
    case onChange
    case passwordChanged
    case firstNameChanged
    case emailChanged
}
```

```
public let emailRegistrationReducer = Reducer<
    EmailRegistrationState,
    EmailRegistrationAction,
    EmailRegistrationEnvironment
> { state, action, env in
    switch action {
                 .... 
                       }
   }
.binding()
```
